x-pack/plugin/core/src/main/java/org/elasticsearch/license/internal/TrialLicenseVersion.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.license.internal;
+
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.xcontent.ToXContentFragment;
+import org.elasticsearch.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.Objects;
+
+/**
+ * Sometimes we release a version with a bunch of cool new features, and we want people to be able to start a new trial license in a cluster
+ * that's already used a trial and let it expire. This class controls when we do that. The serialization of this class is designed to
+ * maintain compatibility with old-school Elasticsearch versions (specifically the {@link org.elasticsearch.Version} class).
+ */
+public class TrialLicenseVersion implements ToXContentFragment, Writeable {
+
+    // Increment this when we want users to be able to start a new trial. Note that BWC with old versions of Elasticsearch cause this to be
+    // limited to a maximum of 99 (inclusive). See the Version class for more info.
+    public static final TrialLicenseVersion CURRENT = new TrialLicenseVersion(8);
+
+    private final int era;
+
+    public TrialLicenseVersion(int era) {
+        this.era = era;
+    }
+
+    public TrialLicenseVersion(StreamInput in) throws IOException {
+        this.era = (in.readVInt() - 99) / 1000000; // Copied from the constructor of Version
+    }
+
+    public static TrialLicenseVersion fromString(String from) {
+        String[] parts = from.split("[.-]");
+        if (parts.length == 0) {
+            throw new IllegalArgumentException("illegal trial era format: [" + from + "]");
+        }
+        try {
+            final int era = Integer.parseInt(parts[0]);
+            return new TrialLicenseVersion(era);
+        } catch (NumberFormatException e) {
+            throw new IllegalArgumentException("unable to parse trial era: [" + from + "]", e);
+        }
+    }
+
+    int asInt() {
+        return era;
+    }
+
+    public boolean ableToStartNewTrialSince(TrialLicenseVersion since) {
+        return since.asInt() < era;
+    }
+
+    @Override
+    public String toString() {
+        return Integer.toString(era);
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        return builder.value(asVersionString()); // suffix added for BWC
+    }
+
+    // pkg-private for testing
+    String asVersionString() {
+        return this + ".0.0";
```

Comment:
Can we use this bwc format only in bwc cases? Does the license maintain the version bit as a string or version id?

Reply:
A string, unfortunately:
https://github.com/elastic/elasticsearch/blob/878127ce7732e2ddab7c06fda375f3ef24b327f2/x-pack/plugin/core/src/main/java/org/elasticsearch/license/LicensesMetadata.java#L155
(⬆️ code is from `main`, to be clear).

Note this is `LicensesMetadata`, rather than in the license itself, so we only store one of these for the whole cluster. We can probably switch to emitting an integer, are there are any cases where we'd need an old version of ES to be able to parse XContent emitted by a newer version? The main case I'm aware of where we need to read a `LicensesMetadata` XContent is restoring a snapshot, and you can't restore a snapshot taken by a newer version than you're currently on AFAIK.

Reply:
As long as we only upgrade the license metadata format after the min version of the cluster has stabilized in the upgrade, we should be ok.